File: coxswain/src/coxswain/ids.py
"""Identifier minting per spec §2.1.

Field names and formats here are normative: ``turn_id`` is
``cx-<epoch_ms>-<6 chars base36>`` and is minted once per user command
submission, before any parse or grounding work starts. ``fingerprint_id`` and
``override_id`` are turn-scoped composite keys of the form
``<turn_id>:<gate_seq>:<fp|ovr>``.

This module is dependency-free (NFR-2) and CPython-importable with no browser
bindings (NFR-1).
"""

from __future__ import annotations

import secrets
import time
from typing import Final

_BASE36_ALPHABET: Final[str] = "0123456789abcdefghijklmnopqrstuvwxyz"
_TURN_ID_RANDOM_CHARS: Final[int] = 6


def _random_base36(length: int) -> str:
    return "".join(secrets.choice(_BASE36_ALPHABET) for _ in range(length))


def mint_turn_id() -> str:
    """Mint a conversation-turn identifier: ``cx-<epoch_ms>-<6 base36 chars>``."""
    epoch_ms = time.time_ns() // 1_000_000
    return f"cx-{epoch_ms}-{_random_base36(_TURN_ID_RANDOM_CHARS)}"


def fingerprint_id_for(turn_id: str, gate_seq: int) -> str:
    """Per-fingerprint-capture identifier: ``<turn_id>:<gate_seq>:fp``."""
    return f"{turn_id}:{gate_seq}:fp"


def override_id_for(turn_id: str, gate_seq: int) -> str:
    """Per-override-use identifier: ``<turn_id>:<gate_seq>:ovr``."""
    return f"{turn_id}:{gate_seq}:ovr"

File: web-repl/shell/coxswain/ids.js
// ids.js -- browser-side mirror of coxswain/src/coxswain/ids.py (§2.1).
//
// Field names and formats are normative: turn_id is
// `cx-<epoch_ms>-<6 chars base36>` and is minted ONCE per user command
// submission, at input capture in coxswain-shell.js, BEFORE any parse or
// grounding work starts (§2.2 point 1). The parse worker and the kernel echo
// it unchanged; nothing downstream ever mints one.
//
// DOM-free (NFR-3), zero dependencies, unit-testable with bun test. Randomness
// is injectable so tests are deterministic; the default path uses
// crypto.getRandomValues when available and falls back to Math.random only as
// a last resort (RISK-12 treats collision risk seriously; epoch-ms plus six
// base36 chars is the specified budget).

const BASE36_ALPHABET = "0123456789abcdefghijklmnopqrstuvwxyz";
export const TURN_ID_RANDOM_CHARS = 6;

function randomBase36(length, randomChar) {
  if (randomChar) {
    let out = "";
    for (let i = 0; i < length; i += 1) out += randomChar();
    return out;
  }
  const cryptoObj = typeof globalThis.crypto !== "undefined" ? globalThis.crypto : null;
  if (cryptoObj && typeof cryptoObj.getRandomValues === "function") {
    const bytes = new Uint8Array(length);
    cryptoObj.getRandomValues(bytes);
    let out = "";
    for (let i = 0; i < length; i += 1) out += BASE36_ALPHABET[bytes[i] % 36];
    return out;
  }
  let out = "";
  for (let i = 0; i < length; i += 1) {
    out += BASE36_ALPHABET[Math.floor(Math.random() * 36)];
  }
  return out;
}

/** Mint a conversation-turn identifier: `cx-<epoch_ms>-<6 base36 chars>`. */
export function mintTurnId({ nowMs = null, randomChar = null } = {}) {
  const epochMs = nowMs === null ? Date.now() : nowMs;
  return `cx-${epochMs}-${randomBase36(TURN_ID_RANDOM_CHARS, randomChar)}`;
}

/** Tab-scoped session id (§2.1), minted once per tab session at panel init. */
export function mintSessionId({ nowMs = null, randomChar = null } = {}) {
  const epochMs = nowMs === null ? Date.now() : nowMs;
  return `cx-sess-${epochMs}-${randomBase36(8, randomChar)}`;
}

/** Per-fingerprint-capture id: `<turn_id>:<gate_seq>:fp` (§2.1). */
export function fingerprintIdFor(turnId, gateSeq) {
  return `${turnId}:${gateSeq}:fp`;
}

/** Per-override-use id: `<turn_id>:<gate_seq>:ovr` (§2.1). */
export function overrideIdFor(turnId, gateSeq) {
  return `${turnId}:${gateSeq}:ovr`;
}
